### PR TITLE
refactor(loader): v2.0.0

### PR DESCRIPTION
**Pull request by d3viant0ne**
##### Compatibility issues

- [ ] `emitFile` error

##### Commit message body
```
 - Enforce options schema
 - Updates code style to Prettier
 - Updates engines
 - Updates peerDependencies
 - Migrates to CircleCI 2.0

BREAKING CHANGE: Drops support for NodeJS 4.x

BREAKING CHANGE: Drops support for Webpack 2.x

```

### CODE REVIEW

File: test/helpers/compiler.js [hunk 1]
```diff
@@ -8,6 +8,8 @@ import path from 'path';
 import webpack from 'webpack';
 import MemoryFS from 'memory-fs';
 
+const majorVersion = require('webpack/package.json').version.split('.')[0];
```
Comment (michael-ciniawsky) [inline test/helpers/compiler.js]: `majorVersion` => `version`

Do we need `mode` for testing ?

File: src/index.js [hunk 2]
```diff
@@ -4,13 +4,14 @@ import validateOptions from 'schema-utils';
...
-  if (!this.emitFile) throw new Error('File Loader\n\nemitFile is required from module system');
+  if (!this.emitFile)
+    throw new Error('File Loader\n\nemitFile is required from module system');
 
   const options = loaderUtils.getOptions(this) || {};
 
   validateOptions(schema, options, 'File Loader');
 
-  const context = options.context || this.rootContext || this.options && this.options.context
+  const context = this.rootContext;
```
Comment (michael-ciniawsky) [inline src/index.js]: `options.context || this.rootContext`

File: test/helpers/compiler.js (hunk 1 above)
Comment (d3viant0ne) [reply test/helpers/compiler.js]: Unless you want to not ever be able to run the test suite on 3.x & 4.x

Comment (d3viant0ne) [reply test/helpers/compiler.js]: `majorVersion` => `version` : It is exactly what it's doing. It's not checking the version, it's grabbing the number that represents the major version.

**Comment by TheLarkInn** [comment]
Looks like just a few conflicts to resolve I can do this if @d3viant0ne you're swamped.

**Comment by michael-ciniawsky** [comment]
There is no urgency here, since the `file-loader =< v1.0.0` works perfectly fine with `webpack >= v4.0.0`, while [this](https://github.com/webpack-contrib/file-loader/pull/235/files#diff-1fdf421c05c1140f6d71444ea2b27638R14) line would  make the loader require `webpack >= v4.0.0`, maybe better to rebase/refactor and just bring `webpack-defaults >= v2.0.0` in for now

**Comment by swernerx** [comment]
Any further progress here? Also lots of open issues and PR without any significant progress. Would wish this PR gets merged to offer follow up PRs for other new features.

**Comment by evilebottnawi** [comment]
@swernerx what problem you have?

**Comment by swernerx** [comment]
I would like to work on making the hashing more flexible e.g. using metrohash and base62. Therefor it would be helpful to have the latest major refactorings merged.

**Comment by evilebottnawi** [comment]
@swernerx here PR don't contain new feature, just refactor code base

**Pull request closed without merging**